pruner/pruner.go

```diff
@@ -0,0 +1,339 @@
+	lock sync.RWMutex
+}
+
+var defaultRebuildTrigger = newTooMuchFiltering(0.2, 1000)
+
+// tooMuchFiltering is the standard rebuildTrigger, which will fire
+// when:
+//
+//   MinAction is less than the sum of counts of found and filtered
+//   patterns and
+//
+//   FilteredToEmitted is greater than the ratio of counts of filtered
+//   and found patterns.
+//
+// defaultRebuildTrigger provides the default trigger policy used by
+// NewMatcher.
+type tooMuchFiltering struct {
+	FilteredToEmitted float64
+	MinAction         int64
+}
+
+func newTooMuchFiltering(ratio float64, min int64) *tooMuchFiltering {
+	return &tooMuchFiltering{
+		FilteredToEmitted: ratio,
+		MinAction:         min,
+	}
+}
+
+func (t *tooMuchFiltering) Rebuild(added bool, s *Stats) bool {
+	if added {
+		// No need to think when we're adding a pattern since
+		// that operation cannot result in an increase of
+		// filtered patterns -- I think.
+		return false
+	}
+
+	if s.Emitted+s.Filtered < t.MinAction {
+		return false
+	}
+
+	if s.Emitted == 0 {
+		return true
+	}
+
+	var (
+		numerator   = float64(s.Filtered)
+		denominator = float64(s.Emitted)
+		ratio       = numerator / denominator
+	)
+
+	return t.FilteredToEmitted < ratio
+}
+
+// DisableRebuild will prevent any automatic rebuilds.
+func (m *Matcher) DisableRebuild() {
+	m.lock.Lock()
+	m.rebuildTrigger = nil
+	m.lock.Unlock()
+}
+
+// rebuildTrigger provides a way to control when rebuilds are
+// automatically triggered during standard operations.
+//
+// Currently an AddPattern, DelPattern, or MatchesForJSONEvent can
+// trigger a rebuild.  When a rebuild is triggered, it's executed
+// synchronous, the the Add/Del method doesn't return until the
+// rebuild is complete.
+type rebuildTrigger interface {
+	// Rebuild should return true to trigger a rebuild.
+	//
+	// This method is called by AddPattern and DelPattern.  added
+	// is true when called by AddPattern; false otherwise. These
+	// methods do not return until the rebuild is complete, so
+	// beware.
+	Rebuild(added bool, s *Stats) bool
+}
+
+// NewMatcher does what you'd expect.
+//
+// The State defaults to MemState.
+func NewMatcher(s State) *Matcher {
+	if s == nil {
+		s = NewMemState()
+	}
+	return &Matcher{
+		Matcher:        quamina.NewMatcher(),
+		live:           s,
+		rebuildTrigger: defaultRebuildTrigger,
+	}
+}
+
+// maybeRebuild calls rebuildTrigger (if not) and then calls rebuild()
+// if that trigger said to do that.  If rebuildTrigger is nil, no
+// rebuild is executed.
+//
+// This method assumes the caller has a write lock.
+func (m *Matcher) maybeRebuild(added bool) error {
+	if m.rebuildTrigger == nil {
+		return nil
+	}
+	if m.rebuildTrigger.Rebuild(added, &m.stats) {
+		return m.rebuild(added)
+	}
+
+	return nil
+}
+
+// AddPattern calls the underlying AddPattern method and then maybe
+// rebuilds the index (if the AddPattern succeeded).
+func (m *Matcher) AddPattern(x quamina.X, pat string) error {
```

Comment:
could we waste 4 bytes and say "pattern"?

Reply:
Trying to stay with Go style that prefers very [short variable names](https://github.com/golang/go/wiki/CodeReviewComments#variable-names). But sure.

I missed this comment in my first pass.